matlab_simulation/04-sensor/get_sonar_range.m

```diff
@@ -0,0 +1,58 @@
+function r_m = get_sonar_range(map, x, fov, r_max)
+% Finds the closest occupied cell within the fov and max range and returns 
+% the inverse measurement model for the sonar sensor. Uses windowing to
+% reduce checked cells.
+%
+% Input:
+%   [map] = True map of environment
+%   [x] = Robot state vector [x position; y position; heading]
+%   fov = Sonar field of view
+%   r_max = Max range of sonar
+% Output:
+%   r_m = Range of object
+
+% Bound the update window within the map dimensions
+[M, N] = size(map);
+w_Mi = max(1,min(M,round(x(1) - r_max)));
+w_Mf = max(1,min(M,round(x(1) + r_max)));
+w_Ni = max(1,min(N,round(x(2) - r_max)));  
+w_Nf = max(1,min(N,round(x(2) + r_max)));  
+w_M = w_Mf - w_Mi + 1;
+w_N = w_Nf - w_Ni + 1;
+win_pos(1) = x(1) - w_Mi + 1;
+win_pos(2) = x(2) - w_Ni + 1;
+
+% FOV limits
+phi_min = -fov/2;
+phi_max = fov/2; 
+
+r_m = r_max;
+
+% Range finder inverse measurement model
+for i = 1:w_M
+    for j = 1:w_N
+        % Map coordinates of window loop cell
+        x_wi = x(1) - win_pos(1) + i;
+        y_wj = x(2) - win_pos(2) + j;
+        
+        % Check if cell is occupied 
+        if (map(x_wi, y_wj) == 0)
+            % Check if edge map (edge of map is considered occupied)
+            if ~(x_wi == 1 || x_wi == M || y_wj == 1 || y_wj == N)
+                continue
```

Comment:
You can avoid having an extra check in your code if you pad all the outer cells with zeros. Then this loop becomes one single if statement. http://www.mathworks.com/help/images/ref/padarray.html

I did this in my pull request. #35 if you want an example. Generally minimizing the number of if statements in a loop greatly increases the code run time in matlab.